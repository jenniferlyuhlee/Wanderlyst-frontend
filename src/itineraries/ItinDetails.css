.Itin-details{
    padding: 1rem 4rem 4rem 4rem;
    margin: 2rem 5rem;
}

@media (max-width: 900px){
    .Itin-details{
        padding: 1rem 3rem 3rem 3rem;
        margin: 2rem;
    }
}

.Itin-info{
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.Itin-likes{
    display: flex;
    align-items: center;
}

.likes-display{
    cursor: auto;
}

.Itin-info-head h3, .Itin-info-head h5 {
    margin: 0;
}

.Itin-tags{
    margin: 0.5rem 0;
}

.Itin-map{
    margin: 0 auto;
    height: 70vh;
    width: 80%;
}


@media (max-width: 1200px){
    .Itin-map{
        height: 60vh;
        width: 100%;
    }
}

@media (max-width: 600px){
    .Itin-map{
        height: 50vh;
    }
}

.Place{
    display: flex;
    flex-direction: column;
}
.Place-img{
    width: 60vw;
    align-self: center;
    margin: 1rem;
}

.bi-trash3-fill{
    font-size: 2.5rem;
}

.bi-trash3-fill:hover{
    cursor: pointer;
}
