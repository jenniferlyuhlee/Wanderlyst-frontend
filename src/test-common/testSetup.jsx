import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };