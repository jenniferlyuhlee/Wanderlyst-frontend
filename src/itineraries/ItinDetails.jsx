import React, { useState, useEffect, useContext, useRef, forwardRef} from "react";
import { useParams, NavLink, useNavigate } from "react-router-dom";
import WanderlystApi from "../utils/api";
import UserContext from "../auth-user/UserContext";
import GoogleMap from "./GoogleMap";
import Loading from "../shared/Loading";
import Alert from "../shared/Alert";
import "./ItinDetails.css"
import { APIProvider} from "@vis.gl/react-google-maps";


function ItinDetails(){
    const { currUser, likes, toggleLike, hasLikedItin } = useContext(UserContext);
    const { id } = useParams();
    const navigate = useNavigate();
 
    // state set to null to use loading spinner
    const [itinerary, setItinerary] = useState(null);
    // error state for invalid param - itinerary id 
    const [error, setError] = useState(false);
    // error message state 
    const [errorMsg, setErrorMsg] = useState(null);
    // status to set liked status - convert id param to integer
    const [liked, setLiked] = useState(hasLikedItin(+id));


    // fetches itinerary details from API upon initial load
    useEffect(function getItinDetails() {
        async function getItin(){
            try{
                setItinerary(await WanderlystApi.getItin(id));
            }
            catch(err){
                console.error("Itinerary not found");
                setError(true);
            }
        }
        getItin();
    }, [id, likes]);

    // function to delete itinerary
    async function deleteItin(){
        try{
           const resp = await WanderlystApi.deleteItin(id, currUser.username);
           if (resp.deleted === +id) navigate('/');
        }
        catch(err){
            setErrorMsg([err]);
        }
    }

    // function to handle likes
    async function handleLike(){
        toggleLike(+id);
        setLiked(likeState => !likeState);
    }

    if(error) return <p className="NotFound"><i>Sorry this itinerary doesn't exist.</i></p>
    if(!itinerary) return <Loading />

    return(
        <div className = "Itin-details card body-cont">
                {errorMsg?
                    <Alert type = "danger" messages = {errorMsg}/>
                    :
                    null
                }
            <div className = "Itin-info mb-2">
                <div className="Itin-info-head">
                <h1>{itinerary.title}</h1>
                    <h3 className="">{itinerary.city}, {itinerary.country}</h3>
                    <h5>{itinerary.duration} days</h5>
                    <div className="Itin-tags">
                    {itinerary.tags?
                        itinerary.tags.map(t => (
                            <NavLink key={t} to ={`/tags/${t}`}>
                                <span className = "badge text-bg-primary">
                                    {t}
                                </span>
                            </NavLink>

                        ))
                        :
                        null
                    }
                    </div>
                    <p>By: <NavLink to={`/users/profile/${itinerary.username}`}>
                            @{itinerary.username}
                        </NavLink>
                    </p>
                    <p>Posted: {itinerary.createdAt.slice(0, 10)}</p>
                </div>
                <div className="Itin-info-likes">
                    {itinerary.username === currUser.username?
                        <>
                        <div className="text-start">
                            <i className="bi bi-trash3-fill" onClick={deleteItin}/>
                        </div>
                        <div className="Itin-likes">
                            <i className ="bi bi-heart likes-display" />
                            <p className="ms-1">{itinerary.likes}</p>
                        </div>
                        </>
                        :
                        <div className="Itin-likes">
                            {liked ?
                                <i className ="bi bi-heart-fill" onClick = {handleLike}/>
                                :
                                <i className="bi bi-heart" onClick = {handleLike}/>
                            }
                            <p className="ms-1">{itinerary.likes}</p>
                        </div>
                    }

                </div>
               
            </div>
            <hr/>
            <p className="mt-2 mb-3">{itinerary.description}</p>
 
            <div className = "Itin-map mb-4">
                <APIProvider apiKey={import.meta.env.VITE_GOOGLE_API_KEY}>
                    <GoogleMap itinerary={itinerary}/>
                </APIProvider>
            </div>

            <div className="Itin-Places">
                    {itinerary.places.map(p => (
                        <div key = {p.seq} className="Place">
                        <p>
                            <b>{p.seq}. {p.name}</b>
                            <br/>
                            <i>{p.address}</i>
                            <br/>
                        </p>
                        <p className="mx-3">{p.description}</p>
                        
                        {!(p.image ==="" || p.image=== null)? 
                            <img src={p.image} alt={p.name} className="Place-img"/> 
                            : 
                            null
                        }
                        </div>
                    ))}
            </div>
        </div>
    )
}

export default forwardRef(ItinDetails);