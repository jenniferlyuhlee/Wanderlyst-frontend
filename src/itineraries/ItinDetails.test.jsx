import React from 'react';
import { fireEvent, render, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter } from 'react-router-dom';
import WanderlystApi from '../utils/api';
import ItinDetails from './ItinDetails';
import { beforeEach, afterEach, vi, expect, describe, test  } from 'vitest';

// Mock the GoogleMap component
describe("ItinDetails component", () => {
    //mock toggleLike
    const mockToggleLike = vi.fn().mockResolvedValue(true);

    // mock google map component
    vi.mock('./GoogleMap', () => ({
        __esModule: true,
        default: () => <div data-testid="mock-google-map">Mock Google Map</div>
    }));

    // Mock the APIProvider to avoid using real API keys
    vi.mock('@vis.gl/react-google-maps', () => ({
        __esModule: true,
        APIProvider: ({ children }) => <div>{children}</div>
     }));

    beforeEach(() => {
        // mock api method getItin returning test data
        vi.spyOn(WanderlystApi, "getItin").mockResolvedValue({
            id: 1,
            username: "testuser",
            title: "test title",
            city: "test city",
            country: "test country",
            duration: "1",
            description: "test desc",
            createdAt: "2024-10-01",
            likes: 1,
            places: [
                    {
                     address: "test address",
                     description: "test place desc",
                     name: "test place",
                     seq: 1,
                    },
                 ],
            tags:  ["testTag"]
        });

        // mock api method deleteItin
        vi.spyOn(WanderlystApi, "deleteItin").mockResolvedValue("deleted");

    });
    afterEach(() => {
        vi.restoreAllMocks();
    });

    test("renders without crashing", async () => {
       await waitFor(() => {render(
            <MemoryRouter initialEntries={['/itineraries/1']}>
                <UserProvider hasLikedItin={() => true}>
                    <ItinDetails/>
                </UserProvider>
            </MemoryRouter>)
        });
    });

    test("matches snapshot", async () => {
        const {getByText, asFragment, getByTestId} = render(
            <MemoryRouter initialEntries={['/itineraries/1']}>
                <UserProvider hasLikedItin={() => true}>
                    <ItinDetails/>
                </UserProvider>
            </MemoryRouter>);

        // wait and check that itinerary is displayed
        await waitFor(() => {
            expect(getByText("test title")).toBeInTheDocument();
            expect(getByTestId("mock-google-map")).toBeInTheDocument();
        });
        expect(asFragment()).toMatchSnapshot();
    });

    test("user can toggle like itinerary", async () => {
        const {getByText, container} = render(
            <MemoryRouter initialEntries={['/itineraries/1']}>
                <UserProvider hasLikedItin={() => false} toggleLike={mockToggleLike}>
                    <ItinDetails/>
                </UserProvider>
            </MemoryRouter>);

        // wait and check that itinerary is displayed
        await waitFor(() => {
            expect(getByText("test title")).toBeInTheDocument();
        });
        
        // click heart and checks that it shows as "liked"
        const likeBtn = container.querySelector(".bi-heart");
        fireEvent.click(likeBtn);
        expect(container.querySelector(".bi-heart-fill")).toBeInTheDocument;
    });

    test("user can delete itinerary", async () => {
        const {getByText, container} = render(
            <MemoryRouter initialEntries={['/itineraries/1']}>
                <UserProvider currUser = {{username: "testuser"}}
                              hasLikedItin={() => false} 
                              toggleLike={mockToggleLike}>
                    <ItinDetails/>
                </UserProvider>
            </MemoryRouter>);

        // wait and check that itinerary is displayed
        await waitFor(() => {
            expect(getByText("test city, test country")).toBeInTheDocument();
        });
        
        // click trash and checks that deleteItin was called
        const deleteBtn = container.querySelector(".bi-trash3-fill");
        fireEvent.click(deleteBtn);
        expect(WanderlystApi.deleteItin).toBeCalled();
    });
});
