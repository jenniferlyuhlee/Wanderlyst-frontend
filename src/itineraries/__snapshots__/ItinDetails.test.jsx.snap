// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItinDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Itin-details card body-cont"
  >
    <div
      class="Itin-info mb-2"
    >
      <div
        class="Itin-info-head"
      >
        <h1>
          test title
        </h1>
        <h3
          class=""
        >
          test city, test country
        </h3>
        <h5>
          1 days
        </h5>
        <div
          class="Itin-tags"
        >
          <a
            class=""
            href="/tags/testTag"
          >
            <span
              class="badge text-bg-primary"
            >
              testTag
            </span>
          </a>
        </div>
        <p>
          By: 
          <a
            class=""
            href="/users/profile/testuser"
          >
            @testuser
          </a>
        </p>
        <p>
          Posted: 2024-10-01
        </p>
      </div>
      <div
        class="Itin-info-likes"
      >
        <div
          class="Itin-likes"
        >
          <i
            class="bi bi-heart-fill"
          />
          <p
            class="ms-1"
          >
            1
          </p>
        </div>
      </div>
    </div>
    <hr />
    <p
      class="mt-2 mb-3"
    >
      test desc
    </p>
    <div
      class="Itin-map mb-4"
    >
      <div>
        <div
          data-testid="mock-google-map"
        >
          Mock Google Map
        </div>
      </div>
    </div>
    <div
      class="Itin-Places"
    >
      <div
        class="Place"
      >
        <p>
          <b>
            1. test place
          </b>
          <br />
          <i>
            test address
          </i>
          <br />
        </p>
        <p
          class="mx-3"
        >
          test place desc
        </p>
        <img
          alt="test place"
          class="Place-img"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
